VoodooRMI/Transports/RMII2C.cpp

```diff
@@ -0,0 +1,286 @@
+        IOLog("%s: Could not open command gate\n", getName());
+        goto exit;
+    }
+
+    /* Implementation of polling */
+    interrupt_source = IOInterruptEventSource::interruptEventSource(this, OSMemberFunctionCast(IOInterruptEventAction, this, &RMII2C::interruptOccured), device_nub, 0);
+    if (!interrupt_source) {
+        IOLog("%s: Could not get interrupt event source\n, trying to fallback on polling.", getName());
+        interrupt_simulator = IOTimerEventSource::timerEventSource(this, OSMemberFunctionCast(IOTimerEventSource::Action, this, &RMII2C::simulateInterrupt));
+        if (!interrupt_simulator) {
+            IOLog("%s: Could not get timer event source\n", getName());
+            goto exit;
+        }
+        work_loop->addEventSource(interrupt_simulator);
+        interrupt_simulator->setTimeoutMS(200);
+        IOLog("%s: Polling mode initialisation succeeded.", getName());
+    } else {
+        work_loop->addEventSource(interrupt_source);
+        interrupt_source->enable();
+        IOLog("%s: running on interrupt mode.", getName());
+    }
+
+    client = getClient();
+    if (!client) {
+        IOLog("%s: Could not get client\n", getName());
+        goto exit;
+    }
+
+    setProperty("VoodooI2CServices Supported", kOSBooleanTrue);
+    reading = false;
+    return res;
+exit:
+    releaseResources();
+    return false;
+}
+
+void RMII2C::releaseResources() {
+    if (command_gate) {
+        command_gate->disable();
+        work_loop->removeEventSource(command_gate);
+    }
+    if (interrupt_source) {
+        interrupt_source->disable();
+        work_loop->removeEventSource(interrupt_source);
+    }
+    if (interrupt_simulator) {
+        interrupt_simulator->disable();
+        work_loop->removeEventSource(interrupt_simulator);
+    }
+    
+    OSSafeReleaseNULL(command_gate);
+    OSSafeReleaseNULL(interrupt_simulator);
+    OSSafeReleaseNULL(interrupt_source);
+    OSSafeReleaseNULL(work_loop);
+    
+    if (device_nub) {
+        if (device_nub->isOpen(this))
+            device_nub->close(this);
+        device_nub->release();
+        device_nub = NULL;
+    }
+    
+    IOLockFree(page_mutex);
+}
+
+void RMII2C::stop(IOService *device) {
+    releaseResources();
+    super::stop(device);
+}
+
+int RMII2C::rmi_set_page(u8 page)
+{
+    uint8_t writeReport[25] = {0x25, 0x00, 0x17, 0x00};
+
+    writeReport[4] = RMI_WRITE_REPORT_ID;
+    writeReport[5] = 1;
+    writeReport[6] = RMI_PAGE_SELECT_REGISTER;
+    writeReport[7] = page;
+    
+    if (device_nub->writeI2C(writeReport, sizeof(writeReport)) != kIOReturnSuccess) {
+        IOLog("%s: failed to write request output report\n", getName());
+        return -1;
+    }
+
+    this->page = page;
+    return 0;
+}
+
+int RMII2C::rmi_set_mode(u8 mode) {
+    u8 command[] = { 0x22, 0x00, 0x3f, 0x03, 0x0f, 0x23, 0x00, 0x04, 0x00, RMI_SET_RMI_MODE_REPORT_ID, mode }; //magic bytes from Linux
```

Comment:
This has a lot of magic numbers, especially when comparing to the Linux source for rmi_i2c. Can we reduce this, or better yet, more closely follows what is under [rmi_i2c](https://github.com/torvalds/linux/blob/master/drivers/input/rmi4/rmi_i2c.c)?

Reply:
I forgot to mention, this part is from [VoodooI2CSynaptics](https://github.com/VoodooI2C/VoodooI2CSynaptics/blob/e8923aa32b15c93af48bcd5f0c1dbdc14d31b228/VoodooI2CSynaptics/VoodooI2CSynapticsDevice.cpp#L899). I haven't look into it yet, just took a quick search but couldn't found it in Linux source code. However, it seems like this part is like reset() to get rat scanning and further reading through,

Reply:
Found it, actually in [hid-rmi.c](https://github.com/torvalds/linux/blob/ffc1c20c46f74e24c3f03147688b4af6e429654a/drivers/hid/hid-rmi.c#L143)

Reply:
Interesting, looks like another layer in addition to RMI4's i2c stuff.

Reply:
Just spent some time jumping across linux sources.
    u8 command[] = {
        0x22, // registerIndex : wCommandRegister
        0x00, // registerIndex+1
        0x3f, // reportID | reportType << 4; reportType: 0x03 for HID_FEATURE_REPORT (kIOHIDReportTypeFeature) ; 0x02 for HID_OUTPUT_REPORT (kIOHIDReportTypeOutput)
        0x03, // hid_set_report_cmd =    { I2C_HID_CMD(0x03) };
        0x0f, // report_id -> RMI_SET_RMI_MODE_REPORT_ID
        0x23, // dataRegister & 0xFF; wDataRegister
        0x00, // dataRegister >> 8;
        0x04, // size & 0xFF; 2 + reportID + buf (reportID excluded)
        0x00, // size >> 8;
        RMI_SET_RMI_MODE_REPORT_ID, // report_id = buf[0];
        mode };
Maybe just left it for now? Functions involved seems not implemented or out of scope of this driver.